### PR TITLE
#25 Integrated bitcoin service with bitcoin testnet

### PR DESCRIPTION
**Pull request by mudiali**
These changes ensure that the bitcoin service runs successfully against testnet after our proof of fee changes. Issue #337 is also being fixed as a result of this work.

### CODE REVIEW

File: lib/bitcoin/BitcoinProcessor.ts [hunk 1]
```diff
@@ -835,7 +864,7 @@ export default class BitcoinProcessor {
     const requestString = JSON.stringify(request);
     // console.debug(`Fetching ${fullPath}`);
     // console.debug(requestString);
-    console.debug(`Sending jRPC request: id: ${request.id}, method: ${request['method']}`);
+    console.debug(`Sending jRPC request: ${JSON.stringify(request)}`);
```
Comment (logangirvin) [inline lib/bitcoin/BitcoinProcessor.ts]: I'd prefer some smarter logic here so we don't end up logging things that would never really be useful like the signed raw transaction on broadcast, but at the very least could this be:
```suggestion
    console.debug(`Sending jRPC request: ${requestString}`);
```

File: lib/bitcoin/BitcoinProcessor.ts [hunk 2]
```diff
@@ -483,10 +474,39 @@ export default class BitcoinProcessor {
...
+    const startingBlockForLastSavedGroupId = this.getStartingBlockFromGroupId(lastSavedGroupId);
+
+    // We want to delete the last group Id because we might not have saved all the transaction(s)
+    // in that group. So removing that ensures that we redo this block which will ensure that
+    // all the transactions are processed again.
+    await this.quantileCalculator.removeGroupsGreaterThanOrEqual(lastSavedGroupId);
+
+    // The starting block is included in the lastGroupId and since we are removing the lastGroupId
+    // we should remove that block from the  the transaction store as well
+    await this.transactionStore.removeTransactionsLaterThan(startingBlockForLastSavedGroupId - 1);
```
Comment (thehenrytsai) [inline lib/bitcoin/BitcoinProcessor.ts]: `removeTransactionsLaterThan` takes in a transaction number, not a block number, so you'll need to construct a transaction number using `TransactionNumber.construct()`

Comment (logangirvin) [reply lib/bitcoin/BitcoinProcessor.ts]:
```suggestion
    await this.transactionStore.removeTransactionsLaterThan(TransactionNumber.construct(startingBlockForLastSavedGroupId - 1, 0));
```

Comment (logangirvin) [reply lib/bitcoin/BitcoinProcessor.ts]:
```suggestion
    await this.transactionStore.removeTransactionsLaterThan(TransactionNumber.construct(startingBlockForLastSavedGroupId, 0) - 1);
```

File: lib/bitcoin/BitcoinProcessor.ts [hunk 3]
```diff
@@ -483,10 +474,39 @@ export default class BitcoinProcessor {
...
 
+  private async getStartingBlockForInitialization (): Promise<IBlockInfo | undefined> {
+    // We look at the latest data that we saved in the quantile calculator. The latest
+    // Id in there tells us which group of blocks have been processed.
+    const lastSavedGroupId = await this.quantileCalculator.getLastGroupId();
+
+    if (!lastSavedGroupId) {
+      // We are returning undefined because at the time of this refactor, the calling code
+      // is expecting an undefined for this case and having this code return the genesis
+      // block is going to cause more refactor.
```
Comment (thehenrytsai) [inline lib/bitcoin/BitcoinProcessor.ts]: Consider adding a new issue to track this and referencing it in comment.

File: lib/bitcoin/BitcoinProcessor.ts (hunk 1 above)
Comment (mudiali) [reply lib/bitcoin/BitcoinProcessor.ts]: Reverting the log message back to the previous one.

File: lib/bitcoin/BitcoinProcessor.ts (hunk 2 above)
Comment (mudiali) [reply lib/bitcoin/BitcoinProcessor.ts]: Updated code to convert block to transaction number and remove it.

File: lib/bitcoin/BitcoinProcessor.ts (hunk 3 above)
Comment (mudiali) [reply lib/bitcoin/BitcoinProcessor.ts]: Updated the comment with the new issue number.

File: lib/bitcoin/BitcoinProcessor.ts [hunk 4]
```diff
@@ -793,7 +800,10 @@ export default class BitcoinProcessor {
       const oneSidetreeTxAlreadyFound = (sidetreeTxToAdd !== undefined);
 
       if (isSidetreeTx && oneSidetreeTxAlreadyFound) {
-        throw new Error('The transaction has more then one sidetree anchor strings.');
+        // tslint:disable-next-line: max-line-length
+        const message = `The outputs in block: ${transactionBlock} with transaction id: ${transactionId} has multiple sidetree transactions. So ignoring this transaction.`;
+        console.debug(message);
+        return false;
```
Comment (thehenrytsai) [inline lib/bitcoin/BitcoinProcessor.ts]: This change looks good, but it raises a question regarding the try-catch block that wraps this method call: should we eat the exception or throw?

Comment (mudiali) [reply lib/bitcoin/BitcoinProcessor.ts]: We shouldn't catch. The catch is there right now only to log debug info.

**Pull request merged**